# postgwas-tools — Workflows

This document provides hands-on examples of how to use the tools included in the `postgwas-tools` package.

## Getting the results from mostest

(From https://github.com/precimed/mostest/blob/master/process_results.py and https://github.com/precimed/mostest/blob/master/process_results_ext.py)

To convert MOSTest results from `.mat` format to text files, use the following commands :
The original scripts were updated to use the `argparse` module instead of manual `sys` argument handling, and to generate a correlation matrix based on the univariate results for each SNP.

#### To get the `.sumstats` file 

```bash
process_results \
  --bim data/reference/UKB26502_QCed_230519_maf0p005_chr21.bim \
  --fname results/all_chr21 \
  --out results/out \
```

**Example out.most_orig.sumstats:**

| CHR | SNP          | BP        | A1 | A2 | PVAL                 | Z_FAKE             | N       |
| --- | ----------   | --------- | -- | -- | -------------------- | ------------------ | ------  |
| 1   | rs12238997   | 693731    | A  | G  | 0.50225630850675     | 0.6709438372570653 | 32964.0 |
| 1   | rs200531508  | 711310    | G  | A  | 0.8217888252639312   | 0.2252448181865746 | 32576.0 |
| 1   | rs570631764  | 713914    | A  | G  | 0.25621477331660036  | 1.1353832532589765 | 33259.0 |
| 1   | rs114983708  | 714019    | A  | G  | 0.26620348259207177  | 1.1118480631888468 | 33202.0 |

#### To get the correlation matrix and the univariate results

```bash
process_results_ext \
  --bim data/reference/UKB26502_QCed_230519_maf0p005_chr21.bim \
  --fname results/all_chr21 \
  --out results/out \
  --uniGWAS
```

If you **omit `--out`**, the script will automatically use `fname` as the output prefix:

```bash
process_results_ext \
  --bim data/reference/UKB26502_QCed_230519_maf0p005_chr21.bim \
  --fname results/all_chr21 \
  --uniGWAS
```

If you **don’t care about `uniGWAS`**, you can skip it completely, when not specifying --uniGWAS:

```bash
process_results_ext \
  --bim data/reference/UKB26502_QCed_230519_maf0p005_chr21.bim \
  --fname results/all_chr21
```

**Example output:**

When using MOSTest, a separate GWAS is performed for each phenotype. Consequently, each SNP is associated with a vector of z-scores — one value per phenotype (e.g., 32 z-scores if there are 32 phenotypes). The following correlation matrix represents the pairwise correlations between these 32-dimensional z-score vectors for each pair of SNPs.

![Correlation matrix Example](docs/images/Corr_matrix_Large_CINGULATE_left.png)

See https://github.com/precimed/mostest?tab=readme-ov-file for more details.

---

## Identify Lead SNPs (FUMA like, with PLINK1.9)

```bash
PLINK=/app/plink

leadSNP --sumstats /data/gwas.sumstats \
        --col-a1 A1 --col-a2 A2 \
        --bfile filt_imputed_autosomes_maf-0.01 \
        --plink $PLINK \
        --out results
```

This command extracts **independent lead SNPs** using FUMA-like logic and produces four files:
`GenomicRiskLoci.txt`, `leadSNPs.txt`, `IndSigSNPs.txt`, and `params.config`.
The first three files follow the same structure as those generated by FUMA.

Internally, the tool runs two PLINK 1.9 commands in sequence.

To get PLINK 1.9 and the command plink, you can do:

```bash
mkdir plink1.9
cd plink1.9
wget -q https://s3.amazonaws.com/plink1-assets/plink_linux_x86_64_20250819.zip 
unzip plink_linux_x86_64_20250819.zip
rm plink_linux_x86_64_20250819.zip
```

### Step 1 – Identify Independent Significant SNPs
```bash
plink --bfile filt_imputed_autosomes_maf-0.01 \
      --clump correct_format_gwas.sumstats \
      --clump-p1 5e-08 \
      --clump-p2 5e-06 \
      --clump-r2 0.6 \
      --clump-kb 1000 \
      --out FUMA/clump_ind
```
This command uses the binary genotype files (`.bed`, `.bim`, `.fam`) from `--bfile` filt_imputed_autosomes_maf-0.01 to calculate linkage disequilibrium (LD) within the sample.
It then **clumps** SNPs from the summary statistics file (`--clump`) as follows:

* Only SNPs with **p-values < 5e-8** (`--clump-p1`) are considered significant.
* SNPs with **p-values < 5e-6** (`--clump-p2`) can be included in clumps.
* SNPs are grouped if **r² ≥ 0.6** (`--clump-r2`), and clumps span at most **1 Mb** (`--clump-kb` 1000).
* Results are stored temporarily in `FUMA/clump_ind`.


Note: FUMA limits clump size by the number of SNPs rather than physical distance;
this implementation differs slightly by limiting by base-pair window size (1 Mb).

### Step 2 – Identify Lead SNPs
```bash
plink --bfile filt_imputed_autosomes_maf-0.01 \
      --clump correct_format_gwas.sumstats \
      --clump-p1 5e-08 \
      --clump-p2 5e-06 \
      --clump-r2 0.1 \
      --clump-kb 1000 \
      --out FUMA/clump_lead
```

This second clumping step repeats the previous logic, except that **r² = 0.1** is used to identify **lead SNPs**.

The output is stored temporarily in `FUMA/clump_lead`.

### Step 3 – Merge clumps

Finally, clumps whose lead SNPs are located within **250 kb** of each other are merged into **genomic loci**.

---

## Fast Manhattan Plot

The **fast Manhattan plot** reduces plotting time by skipping points with p-values above `1e-2`.
Instead of plotting each SNP individually, it draws **rectangles** for each chromosome with:

* **Length** = (last bp position − first bp position)
* **Height** = from 0 to −log₁₀(1e-2)

Annotations from the previous step (e.g., lead SNP names) can be added.
Both `GenomicRiskLoci.txt` and `leadSNPs.txt` are supported.

```bash
manhattan_plot -p /data/gwas.sumstats \
                     --lead-snps GenomicRiskLoci.txt \
                     -o results
```

**Example output:**

![Manhattan Example](docs/images/light_manhattan_example.png)

---

## Multi-Model Manhattan Plot

If you want to plot many results on a same figure to compare different phenotypes, you can.
Also, you can apply a threshold on the maximum `-log10(p-value)` to avoid having a shrinked manhattan plot because of a very low p-value.
To do so, use `--ymax` with the value `40` for instance.
Generate a combined Manhattan plot for multiple GWAS results in a `manhattan_plot.png`:

```bash
manhattan_plot -p /data/gwas1.sumstats /data/gwas2.sumstats /data/gwas3.sumstats \
               -k manhattan \
               -o results
```

Alternatively, you can use a wildcard:
```bash
multi_plot -p "/data/*.sumstats" \
               -k manhattan \
               -o results
```

**Example output:**

![Manhattan Example](docs/images/manhattan_example.png)

If you want to make a Miami plot, it is possible as well. 
Specify the kind with -k as miami to obtain a `miami_plot.png`.

```bash
manhattan_plot -p /data/gwas1.sumstats /data/gwas2.sumstats \
               -k miami \
               --ymax 40 \
               -o results
```

**Example output:**

![Miami Example](docs/images/miami_plot.png)
---

## QQ Plot

Use this command to check GWAS inflation.
Based on code from [ShujiaHuang/qmplot](https://github.com/ShujiaHuang/qmplot):

```bash
QQ_plot -p /data/gwas.sumstats \
        -o results
```

<img src="docs/images/QQplot.png" alt="QQ Example" width="400">

---

## Replication Check by rsID
This tool extracts rows from the replication summary statistics file (`--sumstats`) that correspond to the **lead SNPs** representing genomic loci in the `GenomicRiskLoci.txt` file.  
It matches SNPs based on the `rsID` column.

It can take either `leadSNPs.txt` **or** `GenomicRiskLoci.txt` as input for the `--lead-snps` argument.

```bash
replication_by_rsid \
        --lead-snps GenomicRiskLoci.txt \
        --sumstats /data/replication.sumstats \
        --out replication_results
```
**Example output:**

| CHR | SNP        | BP        | A1 | A2 | PVAL                 | Z_FAKE             | N      |
| --- | ---------- | --------- | -- | -- | -------------------- | ------------------ | ------ |
| 1   | rs12408663 | 19641564  | C  | T  | 0.1722612687472529   | 1.3649738512483902 | 9950.0 |
| 2   | rs6546175  | 65749771  | C  | T  | 0.42544035829526683  | 0.797018377419405  | 9950.0 |
| 3   | rs6778735  | 52531084  | C  | T  | 0.3536629004493488   | 0.9275078836756854 | 9950.0 |
| 6   | rs58321169 | 126547421 | T  | C  | 0.002508548341954122 | 3.0223083972754567 | 9950.0 |

---

## LDSC Results Extraction

This section explains how to extract **heritability estimates** and **genetic correlations** from LDSC (Linkage Disequilibrium Score Regression) results.


To extract SNP-based heritability estimates from `.log` files generated by LDSC and save them to a `.tsv` table:
```bash
extract_h2 -p ldsc/h2/*.log \
           -o results
```

**Example output (`h2_summary.tsv`):**

| pheno | h2     | se     | lambdaGC | Mean_Chi2 | Intercept |
| ----- | ------ | ------ | -------- | --------- | --------- |
| dim18 | 0.1414 | 0.0161 | 1.0895   | 1.1062    | 1.0061    |
| dim13 | 0.1087 | 0.0156 | 1.0649   | 1.0810    | 1.0048    |
| dim15 | 0.0716 | 0.0154 | 1.0496   | 1.0605    | 1.0105    |


To extract LDSC genetic correlation results from `.log` files and save them to a `.tsv` table:


```bash
extract_gencorr -p ldsc/SCZcorr/scz_dim*.log \
                -o results \
                --prefix scz_
```



**Example output (`gencorr_summary.tsv`):**

| pheno | gencov  | gencorr  | se     | zscore | P      |
| ----- | ------- | -------- | ------ | ------ | ------ |
| dim1  |-0.0157  | -0.0778  | 0.0448 | -1.736 | 0.0824 |
| dim2  |0.0158	  | 0.0875   | 0.0452 | 1.933  | 0.0532 |	
| dim3  |0.0047	  | 0.0247   | 0.0423 | 0.5842 | 0.5591 |


And if you want to compute an omnibus test on the different results to combine them in a meta analysis, cou can add:
It will only keep the phenotypes where the h2 can be estimated, to remove noise from the test.
(If a phenotype yields a line such as `Genetic Correlation: nan (nan) (h2  out of bounds)` in the `.log` file, it won't be taken into account within the test.)

```bash
extract_gencorr -p ldsc/scz_dim*.log \
                -o tests \
                --prefix scz_ \
                --omnibus
```

**Example output (`meta_scz_ldsc.json`):**

{
    "chi2_statistic": 7.094600250000001,
    "degrees_of_freedom": 3,
    "p_value": 0.06894289054589708,
    "phenotypes": [
        "dim1",
        "dim2",
        "dim3"
    ]
}

---

## Notes

- All manhattan plots are saved as `.png` files with **transparent backgrounds** for easy embedding.  
- Works with both `.sumstats` and compressed `.gz` files.  
- Regex paths and file lists are both supported.

---

## Back to Overview

Return to the main [README.md](README.md).